pkg/policy/repository.go

```diff
@@ -351,51 +353,121 @@ nextLabel:
 // This is just a helper function for unit testing.
 // TODO: this should be in a test_helpers.go file or something similar
 // so we can clearly delineate what helpers are for testing.
-func (p *Repository) Add(r api.Rule) (uint64, error) {
+func (p *Repository) Add(r api.Rule, localRuleConsumers []IdentityConsumer) (uint64, map[uint16]struct{}, error) {
 	p.Mutex.Lock()
 	defer p.Mutex.Unlock()
 
 	if err := r.Sanitize(); err != nil {
-		return p.revision, err
+		return p.revision, nil, err
 	}
 
 	newList := make([]*api.Rule, 1)
 	newList[0] = &r
-	return p.AddListLocked(newList), nil
+	_, rev := p.AddListLocked(newList)
+	return rev, map[uint16]struct{}{}, nil
+}
+
+// NewIDSet returns a new instance of an IDSet.
+func NewIDSet() *IDSet {
+	return &IDSet{
+		IDs: map[uint16]struct{}{},
+	}
+}
+
+type IdentityConsumer interface {
+	GetID16() uint16
+	GetSecurityIdentity() *identity.Identity
+	PolicyRevisionBumpEvent(rev uint64)
 }
 
 // AddListLocked inserts a rule into the policy repository with the repository already locked
 // Expects that the entire rule list has already been sanitized.
-func (p *Repository) AddListLocked(rules api.Rules) uint64 {
-	newList := make([]*rule, len(rules))
+func (p *Repository) AddListLocked(rules api.Rules) (ruleSlice, uint64) {
+
+	newList := make(ruleSlice, len(rules))
 	for i := range rules {
-		newList[i] = &rule{Rule: *rules[i]}
+		newRule := &rule{
+			Rule:     *rules[i],
+			metadata: newRuleMetadata(),
+		}
+		newList[i] = newRule
 	}
+
 	p.rules = append(p.rules, newList...)
 	p.revision++
 	metrics.PolicyCount.Add(float64(len(newList)))
 	metrics.PolicyRevision.Inc()
 
-	return p.revision
+	return newList, p.revision
+}
+
+// UpdateLocalConsumers updates the cache within each rule in the given repository
+// which specifies whether said rule selects said identity. Returns a wait group
+// which can be used to wait until all rules have had said caches updated.
+func (p *Repository) UpdateLocalConsumers(identifiers []IdentityConsumer) *sync.WaitGroup {
```

Comment:
This is only appropriate for usage during startup, right? because the list of endpoints to update is being ignored.

Reply:
It can be used whenever - it's more of a 'force-update' of the caches within the rules, which could theoretically be called any time for the policy repository. The `IDSet` that is populated is not used anywhere though, although it could be in the future. Would you like me to add an additional function that does not use an `IDSet` as a parameter?

Reply:
If the parameter is ignored, it probably doesn't need to be there (..unless it's satisfying an interface or something). Also, if this is a brute-force "update all" then it's not appropriate for all situations and maybe it'd be good to have a less innocuous-sounding name for the function, just so someone doesn't naively use it thinking it will do a small amount of work.